File: web/tests/reference/recursion-backtracking/generate-parentheses.py
def generate_parenthesis(n):
    result = []

    def backtrack(path, open_count, close_count):
        if len(path) == 2 * n:
            result.append("".join(path))
            return
        if open_count < n:
            path.append("(")
            backtrack(path, open_count + 1, close_count)
            path.pop()
        if close_count < open_count:
            path.append(")")
            backtrack(path, open_count, close_count + 1)
            path.pop()

    backtrack([], 0, 0)
    return result

File: web/tests/reference/recursion-backtracking/generate-parentheses.js
function generateParenthesis(n) {
  const result = [];

  const backtrack = (path, openCount, closeCount) => {
    if (path.length === 2 * n) {
      result.push(path.join(""));
      return;
    }
    if (openCount < n) {
      path.push("(");
      backtrack(path, openCount + 1, closeCount);
      path.pop();
    }
    if (closeCount < openCount) {
      path.push(")");
      backtrack(path, openCount, closeCount + 1);
      path.pop();
    }
  };

  backtrack([], 0, 0);
  return result;
}
